docs/apis/subsystems/roles.md

```diff
@@ -0,0 +1,141 @@
+| `CONTEXT_COURSECAT` | a course category  | 40            |
+| `CONTEXT_COURSE`    | a course           | 50            |
+| `CONTEXT_MODULE`    | an activity module | 70            |
+| `CONTEXT_BLOCK`     | a block            | 80            |
+
+An authorized user will be able to assign an arbitrary number of roles to each user in any context.
+
+See [Roles and modules#Context](https://docs.moodle.org/dev/Roles_and_modules#Context) for more information.
+
+### Capabilities
+
+Capabilities can have the following permissions:
+
+1. `CAP_INHERIT`
+2. `CAP_ALLOW`
+3. `CAP_PREVENT`
+4. `CAP_PROHIBIT`
+
+If no permission is defined, then the capability permission is inherited from a context that is more general than the current context. If we define different permission values for the same capability in different contexts, we say that we are overriding the capability in the more specific context.
+
+### Capability conflicts
+
+Since the capabilities in each role could be different, there could be conflict in capabilities.
+
+If we set a `PROHIBIT` on a capability, it means that the capability cannot be overridden and will ALWAYS  have a permission of prevent (deny). Prohibit always wins.   For example, Jeff has a naughty student role that prohibits him from postings in any forums (for the whole site), but he's also assigned a facilitator role in "Science forum" in the course Science and Math 101. Since prohibit always wins, Jeff is unable to post in "Science forum".
+
+## Programming Interface
+
+Although the Roles system may look complicated at first glance, implementing it in Moodle code is fairly simple.
+
+- Examples of functions handling the **Roles** can be found at [`accesslib.php`](https://github.com/moodle/moodle/blob/master/lib/accesslib.php)
+
+- Moodle comes with a list of predefined roles for example Student, Teacher, Course Manager etc. A new role can be created based on
+    those predefined roles using **role archetype:**
+
+```php
+/**
+ * Returns array of all role archetypes.
+ *
+ * @return array
+ */
+function get_role_archetypes() {
+    return array(
+        'manager'        => 'manager',
+        'coursecreator'  => 'coursecreator',
+        'editingteacher' => 'editingteacher',
+        'teacher'        => 'teacher',
+        'student'        => 'student',
+        'guest'          => 'guest',
+        'user'           => 'user',
+        'frontpage'      => 'frontpage'
+    );
+}
+```
+
+- On each page you need to find the context the user is working in, using the `get_context_instance()` function.  For example, in the forum module:
+
+```php
+$context = get_context_instance(CONTEXT_MODULE, $cm->id);
+```
+
+or at the course level:
+
+```php
+$context = get_context_instance(CONTEXT_COURSE, $id);
+```
+
+- Then, whenever you want to check which role has a given capability do something like this:
+
+```php
+list($roleids) = get_roles_with_cap_in_context($context, 'moodle/course:manageactivities');
+```
```

Comment:
We should discourage this. It's one of the things that people think they need but almost never actually need. It leads to questions like "How do I know who my teachers are", and "How do I make this work for teachers only".

Perhaps put it in a `<details>` section so that it is clear that it is not a part of the norm:

```suggestion
<details>
<summary>Fetching roles and users who hold a capability</summary>

Moodle has a flexible and detailed capability system which allows administrators to define many similar roles for different purposes. It is quite common to have multiple teacher-like roles but need to restrict their access depending on their usage.

For example in a University setting you may have a lecturer who presents the course materials, and then a number of Ph.D students who lead smaller groups of students in labs, seminars, and workshops. These roles may both be considered a form of teacher, but they will have different permissions to suit their needs.

As a result we strongly discourage that you think in terms of which roles or users hold a capability, but rather whether a specific user holds a capability.

:::danger

There are some situations where you _do_ need to get a list of roles with a capability in a specific context, but these are very rare. You can do so using the `get_roles_with_cap_in_context()` function:

```php title="Fetching a list of roles which old a capability in the specified context"
[$roleids] = get_roles_with_cap_in_context($context, 'moodle/course:manageactivities');
```

:::

</detail>
```